Unsafe derives and attributes

joshtriplett opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rfcs/pull/3715)*

Allow declaring proc macro attributes and derive macros as unsafe, and
requiring `unsafe` to invoke them.

[Rendered](https://github.com/joshtriplett/rfcs/blob/unsafe-derives/text/3715-unsafe-derives-and-attrs.md)

joshtriplett committed:
Unsafe derives and attributes

programmerjake commented on text/0000-unsafe-derives-and-attrs.md:
````diff
@@ -0,0 +1,87 @@
+`unsafe`:
+
+```rust
+#[proc_macro_attribute]
````
```suggestion
#[proc_macro_attribute(unsafe)]
```

joshtriplett replied:
Argh. Good catch.

joshtriplett committed:
RFC 3715

ds84182 commented on text/3715-unsafe-derives-and-attrs.md:
````diff
@@ -0,0 +1,87 @@
+```
+
+Invoking this derive requires writing either
+`#[unsafe(derive(DangerousTrait))]` or `#[derive(unsafe(DangerousTrait))]`.
````
Wondering if you've considered unsafe without parenthesis.`#[derive(unsafe DangerousTrait)]` has less visual noise, at least to my eyes.

programmerjake replied:
the `unsafe` uses parenthesis syntax was already decided on in [RFC 3325](https://github.com/rust-lang/rfcs/blob/master/text/3325-unsafe-attributes.md)

kennytm replied:
`unsafe DangerousTrait` may be mistaken as a #3710-unsafe-flavored variant of the `DangerousTrait` trait, so no it shouldn't be considered.

joshtriplett replied:
Added to the alternatives section with that rationale.

joshtriplett committed:
Mention alternative syntax and explain why we don't use it

clarfonthey commented:
Just my 2¢, but I think that the shorthand for `derive(..., unsafe(Trait), ...)` is a bit unprecedented.

You have to separate out derive traits any time there's some different requirement, e.g. `cfg_attr(feature = "serde", derive(Serialize, Deserialize))`, and by keeping unsafe at the top level you can easily grep for `#!?\[unsafe\(` whereas `unsafe(` will catch any call to `method_unsafe(`.

Sure, it's likely it won't make a difference, but I think that only having to check attributes at the top level for unsafe to verify safety is best.

carbotaniuman commented:
The greppability is already broken by things like `#[cfg_attr(all(), unsafe(no_mangle)]`, and can be restored by just grepping `unsafe(` I think putting the unsafe with the trait makes syntactic sense as it discharges the safety obligation of each trait derive.

GnomedDev commented:
Any reason the syntax for defining an unsafe derive/attribute isn't `unsafe fn`?
```rust
#[proc_macro_attribute]
pub unsafe fn dangerous(_attr: TokenStream, item: TokenStream) -> TokenStream {
    item
}
```

Noratrieb commented on text/3715-unsafe-derives-and-attrs.md:
```diff
@@ -0,0 +1,92 @@
+
+## Attributes
+
+When declaring a proc macro attribute, you can add the `unsafe` parameter to
```
@GnomedDev said
> Any reason the syntax for defining an unsafe derive/attribute isn't `unsafe fn`?
```rust
#[proc_macro_attribute]
pub unsafe fn dangerous(_attr: TokenStream, item: TokenStream) -> TokenStream {
    item
}
```

Noratrieb replied:
because that makes no sense, the unsafety is at a meta-level and there are no safety requirements on the macro function itself

GnomedDev replied:
Is there not? I thought the point of an unsafe derive/attribute was that it was unsafe to trigger and had safety requirements?

Noratrieb replied:
you're confusing the memory safety of the rustc process including the proc macro code with the safety of the generated code. this RFC is about the latter, making the macro an unsafe fn is about the former

GnomedDev replied:
I don't personally think it's that confusing, especially for users who don't know how proc macros are loaded and run in the current implementation of the compiler.

clarfonthey commented:
> The greppability is already broken by things like `#[cfg_attr(all(), unsafe(no_mangle)]`, and can be restored by just grepping `unsafe(` I think putting the unsafe with the trait makes syntactic sense as it discharges the safety obligation of each trait derive.

I feel kind of silly for literally alluding to this point in my post and missing it somehow. You're right and I retract my original claim.

joshtriplett committed:
Unresolved question for unsafe placement

joshtriplett commented:
We had a @rust-lang/lang design meeting today on the set of macro RFCs. I've updated the RFC to incorporate the feedback from that design meeting.

Per the feedback in that meeting, I'm starting an FCP to start letting people register consensus for this RFC.

@rfcbot merge

joshtriplett committed:
Resolve the unresolved question about unsafe placement, and document rationale

joshtriplett commented:
@rust-lang/lang I've updated this to match the consensus syntax of `derive(unsafe(UnsafeTrait))`, and provided all the points of rationale that I remember from the discussion. Please feel free to post further suggestions if I missed any points of rationale.

carbotaniuman commented:
One thing I thought of was that it might make sense to make a derive maybe-unsafe and branch between whether `unsafe` was provided, although it might be better to just spell that out as 2 separate derives... The use case for that would be something where the derive can check the preconditions using `static_assert`s or similar, and one where you can promise the preconditions hold if they aren't checkable.

RalfJung commented:
> One thing I thought of was that it might make sense to make a derive maybe-unsafe and branch between whether unsafe was provided

I think that would be a big mistake. The presence or absence of `unsafe` should never affect the resulting generated code, it should only affect whether the code is accepted or not.

RalfJung commented:
> @rust-lang/lang I've updated this to match the consensus syntax of `derive(unsafe(UnsafeTrait))`, and provided all the points of rationale that I remember from the discussion. Please feel free to post further suggestions if I missed any points of rationale.

In terms of the mental model I find `derive(unsafe(Trait))` a bit odd -- this looks like the trait is called `unsafe(Trait)`. In the grammar of attributes (well, in its simplified form in my head), `unsafe($attr)` is also an `$attr`, and `derive($trait)` is an `$attr`, but to fit the proposed scheme into that model we have to do an ad-hoc extension of the grammar to allow `unsafe` inside` derive`.

OTOH, the RFC does make some good points in favor of "inner `unsafe`", so I am a bit torn here.

Veykril commented on text/3715-unsafe-derives-and-attrs.md:
```diff
@@ -0,0 +1,122 @@
+struct SomeStruct { ... }
+
+#[derive(
+    /// SAFETY: ...
```
This strikes me as a bit odd, is this supposed to be a line comment (that is non-doc comment)? Documentation comments only have very specific places where they are allowed. Also they represent `#[doc = "content"]` so this would be equivalent to writing
```rust
#[derive(
    #[doc = "SAFETY: ..."]
    unsafe(DangerousTrait)
...
```

joshtriplett committed:
Fix SAFETY comments to be non-doc comments

joshtriplett committed:
Add another rationale for `derive(unsafe(UnsafeTrait))`, from TC

joshtriplett commented:
@traviscross Thanks for the reminder. IIRC that was one of many possible rationales that came up in the meeting, but it's one that I hadn't captured. I've now done so.

RalfJung commented:
> The argument we landed on, on this point (setting aside the ergonomic arguments in favor of the inner unsafe), is that the derive word is always safe. It's just signaling that we're going to invoke a code transformation. It's the derive macro that is unsafe. That's what has the proof obligation. And that's what the path represents.

I don't understand that distinction. `no_mangle` also just invokes a particular code generation. But that generated code can lead to unsoundness and that's what makes in unsafe. This applies equally to `derive`.

Or put differently, a function invoking an unsafe function without checking its preconditions needs to be unsafe.

But good point that the name in the `derive` is a macro name, not a trait name. That had temporarily escaped my memory.

kennytm commented on text/3715-unsafe-derives-and-attrs.md:
```diff
@@ -0,0 +1,126 @@
+  desired. Putting the `unsafe` on the outside requires separate `derive`s for
+  safe and unsafe derives, and potentially multiple if the derives care about
+  ordering.
+- This makes it easy to attach `SAFETY` comments to each individual trait.
```
Also I don't see how I also don't see how this is an advantage of `#[derive(unsafe(Foo))]` vs `#[unsafe(derive(Foo))]`. The latter is even 2 lines shorter.

```rust
#[derive(
    // SAFETY: ...
    unsafe(DangerousTrait),
    // SAFETY: ...
    unsafe(AnotherDangerousTrait),
)]
struct AnotherStruct { ... }

vs.

// SAFETY: ...
#[unsafe(derive(DangerousTrait))]
// SAFETY: ...
#[unsafe(derive(AnotherDangerousTrait))]
struct AnotherStruct { ... }
```

joshtriplett replied:
@kennytm This goes along with the previous point about not having to split the `derive`.

kennytm commented on text/3715-unsafe-derives-and-attrs.md:
```diff
@@ -0,0 +1,126 @@
+
+Some rationale for putting it on the inside:
+- This encourages minimizing the scope of the `unsafe`, isolating it to a
+  single trait.
```
one could make `#[unsafe(derive(Foo, Bar))]` invalid syntax.

joshtriplett replied:
I don't think we should use a syntax that naturally leads people to assume this would work, and then complain when they do the natural thing that seems like it should work.

Nonetheless, I've captured this in the alternatives section.

kennytm commented on text/3715-unsafe-derives-and-attrs.md:
```diff
@@ -0,0 +1,126 @@
+- This allows writing all traits to derive within a single `#[derive(...)]`, if
+  desired. Putting the `unsafe` on the outside requires separate `derive`s for
+  safe and unsafe derives, and potentially multiple if the derives care about
+  ordering.
```
afaik `#[derive(X)]` cannot see another `#[derive(Y)]` and all a derive-macro does is append an item, which is normally order-agnostic.

> [[items.decl-order]](https://doc.rust-lang.org/reference/items.html#r-items.decl-order)
>
> Items may be defined in **any order**, with the exception of `macro_rules` which has its own scoping behavior.

joshtriplett replied:
@kennytm While I think it's good practice to make derives order-independent, it's very possible to make them order-dependent, in multiple ways. First, as you noted, the are exceptions like macros. Second proc macros can have side effects outside the scope of the generated code.

Nonetheless, I've called out that this is unusual.

kennytm reviewed (commented):
While to the proc_macro author in `#[derive(Clone, Copy)] struct Foo { ... }` it may be understood as invoking the "macros" `Clone` and `Copy` on `struct Foo { ... }`, to an average user it is highly unintuitive to think like that. 

We would think
1. `#[derive(Clone, Copy)]` is just a short hand for `#[derive(Clone)] #[derive(Copy)]` and 
2. the `derive` and the trait (ok, macro) name is a single unit to drive the code generation. 

So `#[unsafe(derive(UnsafeTrait))]` makes more sense to me, which the `unsafe` is applied the entirety of `derive(UnsafeTrait)`, not the word `derive` alone.

carbotaniuman commented on text/3715-unsafe-derives-and-attrs.md:
````diff
@@ -0,0 +1,126 @@
+
+## Derives
+
+When declaring a proc macro `derive`, you can add the `unsafe` parameter to the
+`proc_macro_derive` attribute to indicate that the derive requires `unsafe`:
+
+```rust
+#[proc_macro_derive(DangerousTrait, unsafe)]
+pub fn derive_helper_attr(_item: TokenStream) -> TokenStream {
+    TokenStream::new()
+}
+```
````
On one hand I like the symmetry, on the other hand (because we overloaded unsafe to have two meanings) I think this could be read as the derive macro itself discharging a safety obligation, whereas the first form makes it slightly clearer that this is just an unsafe derive.

joshtriplett replied:
@carbotaniuman That's my feeling as well: `proc_macro_derive(unsafe(DangerousDeriveMacro))` reads to me as though the creation of it is unsafe, rather than it being a safe operation that creates an unsafe derive.

joshtriplett replied:
I've added an alternative for this, along with rationale for why we shouldn't do it.

joshtriplett commented on text/3715-unsafe-derives-and-attrs.md:
```diff
@@ -0,0 +1,126 @@
+Invoking an unsafe attribute requires the unsafe attribute syntax:
+`#[unsafe(dangerous)]`.
+
+If writing code that enforces `SAFETY` comments for every use of `unsafe`, you
+can write the `SAFETY` comment immediately prior to the attribute:
```
@traviscross It is, in a very practical sense, *not* currently the default: there's no warn-by-default lint for it. I'm all for changing that, but meanwhile, it'd be inaccurate to imply that it is currently required everywhere.

joshtriplett committed:
Switch language to reference derive macros rather than traits

Co-authored-by: Travis Cross <tc@traviscross.com>

joshtriplett commented:
@traviscross As mentioned by @carbotaniuman and myself in response to that suggestion: that makes it look like something that's unsafe to *define* rather than unsafe to *use*. For that reason, I don't think that syntax is an improvement.

joshtriplett committed:
More wording updates from "trait" to "derive macro"

joshtriplett committed:
Add alternative for restricting to one derive macro per `unsafe(derive(...))`

joshtriplett committed:
Flag order-dependent derives as an unusual case

joshtriplett committed:
Add alternative for declaring with `proc_macro_derive(unsafe(DangerousDeriveMacro))`

RalfJung commented:
We're already using `unsafe` both to mark the point of introducing unsafe obligations and of discharging them. This is a point that often trips people up (it is on the short list of things I would change about Rust 1.0 if I could). For the existing cases at least there's  fairly short list of you have to learn to know which is which. The proposal to use `unsafe(Trait)` for both cases means that in the future one will have to know exactly how all attributes use `unsafe` to understand which of them introduces vs discharges obligations -- that does not seem great.

programmerjake commented:
it just occurred to me that we may want to mark derive macro helper attributes `unsafe` instead of (or in addition to) the derive macro:
```rust
#[proc_macro_derive(MyDebug, attributes(unsafe(debug_deref)))]
fn my_debug ...
```

```rust
// no `unsafe(MyDebug)` here since it's only unsafe
// when you need to deref pointers, which not all structs have.
#[derive(MyDebug)]
pub struct FooPtr<T> {
    // SAFETY: ptr points to a valid memory chunk allocated with `malloc`
    #[unsafe(debug_deref)]
    ptr: NonNull<T>,
    bar: usize,
}
```

tmandry commented:
The motivation section is really short. Even though this seems like an "obvious" extension, I would appreciate it if @joshtriplett and others could add concrete examples of derives that would use this. Until then further bikeshedding on the spelling seems premature.

joshtriplett commented:
@programmerjake That exact proposal is in the future possibilities section:
> We could provide a syntax to declare specific helper attributes of a derive as unsafe, without declaring the entire derive unsafe.

joshtriplett commented on text/3715-unsafe-derives-and-attrs.md:
```diff
@@ -0,0 +1,142 @@
+- This encourages minimizing the scope of the `unsafe`, isolating it to a
+  single derive macro.
+- This allows writing all derive macros to invoke within a single
+  `#[derive(...)]`, if desired. Putting the `unsafe` on the outside requires
+  separate `derive`s for safe and unsafe derives, and potentially multiple in
+  the (unusual) case where the derives care about ordering.
+- This makes it easy to attach `SAFETY` comments to each individual derive
+  macro.
```
@kennytm That's true, though we could fix that in rustfmt in the future.

joshtriplett replied:
I've added a note about this.

joshtriplett committed:
Rewrite descriptions of SAFETY comments

Use phrasings that can't be read as implying SAFETY comments are
unusual. The previous phrasing could be interpreted as "in the unusual
case where you're writing SAFETY comments ...".

joshtriplett committed:
Add note about rustfmt

joshtriplett committed:
Mention and discuss alternative of unsafe supertrait

joshtriplett commented:
@moulins
> > it just occurred to me that we may want to mark derive macro helper attributes `unsafe` instead of (or in addition to) the derive macro:
> > [...snip]
> 
> I'd go even further: the same attribute could be either safe or `unsafe`, depending on other attributes of the same macro, so there should be a way to accept both `#[unsafe(debug_deref)]` and `#[debug_deref]`.

We could certainly do that, but I think it's reasonable for that to be future work. *This* proposal introduces always-unsafe derives and always-unsafe attributes. In the future, we could have an RFC for sometimes-unsafe derives and sometimes-unsafe attributes, which would require additional definition and documentation.

joshtriplett commented:
@tmandry https://github.com/facet-rs/facet/issues/80 is a good example of rationale for this.

nikomatsakis commented:
@rfcbot reviewed 

I am happy with the syntax and I think you should write `unsafe(Trait)` in both the def and use site. 

I like the idea of unsafe attributes.

joshtriplett committed:
Change syntax to `proc_macro_derive(unsafe(DangerousDeriveMacro))`

joshtriplett committed:
Expand motivation

joshtriplett committed:
Reorder attributes before derives

joshtriplett committed:
Add unsafe helper attributes

tmandry commented:
I still haven't seen a worked example of where and how this will be used. The conclusion of https://github.com/facet-rs/facet/issues/80 was that it was not a good candidate for an unsafe derive.

`Send` and `Sync` were brought up and they seem like decent candidates. (Modulo the detail that they if they were in core they would be built in derives and not proc macros.. but if this concept was useful in core it would probably be useful elsewhere.)

The problem I see is that the ordinary derive logic doesn't make sense for `Send` and `Sync`. If I have

```rust
struct Ref<T>(*const T);
```

Then I most likely want my `Send` impl to look like this:

```rust
unsafe impl<T: Sync> Send for Ref<T> {}
```

We don't have a precedent for a built-in derive that works like this. If we go by the way those derives work today, `#[derive(unsafe(Send))]` would give you

```rust
unsafe impl<T: Send> Send for Ref<T> {}
```

which would be subtly wrong (albeit in a conservative direction), and worse, hidden from the user.

Overall I'm concerned by combination of the following:

* Broad, whole-type-behavior safety invariants implied by the use of a `#[derive(unsafe(...)]`
* The inherent subtlety of unsafe impls, which might require more detailed knowledge of a type's behavior than a derive macro can have
* Opaqueness of derive expansions masking potential problems with the above
* The limited syntax-level knowledge of macros and their inability to see through other types

I think many of these challenges can be overcome, but we should have use cases in mind to guide this and related features. That way we will end up shipping a good user experience (at least for those use cases) instead of a feature that's almost-but-not-quite useful.

tmandry commented:
@rfcbot concern needs fleshed out use case

joshtriplett committed:
Fix typo

joshtriplett commented:
@tmandry A few examples of concrete traits that definitely need unsafe derives *and* don't have any complexity around bounds:

- `TrustedLen` and `TrustedStep`
- `DerefPure`
- [`pyo3::marker::Ungil`](https://docs.rs/pyo3/0.24.2/pyo3/marker/trait.Ungil.html)
- A hypothetical `derive` for [`bytes::Buf`](https://docs.rs/bytes/1.10.1/bytes/buf/trait.Buf.html) or similar. (For cases where your type has fields or trivial expressions for current slice, offset, etc.)

joshtriplett committed:
RFC 3715: Expand motivation

kennytm commented on text/3715-unsafe-derives-and-attrs.md:
```diff
@@ -0,0 +1,188 @@
+  but such a derive would be useful when they're stable, serving the function
+  of an `unsafe impl`.)
+- `pyo3::marker::Ungil` in `pyo3`, in place of the current handling of a
+  blanket impl for any `Send` type.
```
If these are the only motivation examples, I don't get why these *marker* traits `TrustedLen`, `TrustedStep`, `DerefPure`, `Ungil` *require* using the `#[derive]` mechanism over manual `unsafe impl`. 



For the first three the only advantages are not having to explicitly write the `where` bounds, but they have the `Iterator` or `Deref` supertraits which cannot be derived so you gotta repeat those bounds anyway.

joshtriplett replied:
Literally any `unsafe trait` that one might wish to `derive` is a potential example. These were a few samples. If you have other examples of `unsafe trait` that you'd prefer to have cited instead of or in addition to these, I'd be happy to add them.

kennytm replied:
> Literally any `unsafe trait` that one might wish to `derive` is a potential example.

Normally if we wish to `derive` something it is because we know that it is possible to fill out the `impl` programmatically with the type's definition alone, thus saving effort to produce the boilerplate impl. 

If all the derive doing is just implementing a marker trait without anything extra there's certainly no point *from the supply-side* anyone wish to write a proc-macro for it. At least it should be derivable together with its base trait like `#[derive(Clone, Copy)]` i.e. `#[derive(Iterator, unsafe(TrustedLen))]`, but you can't & *can't* derive an Iterator.

Another reason we wish to `derive` a marker trait is to insert some compile-time checks, such as `zerocopy`'s macros. But `zerocopy`'s derive-macro checks at the same proved the conditions required by the `unsafe trait` in the first place, making those traits *safe to derive* even if *unsafe to manually implement*. 

For `libstd` the public, documented, non-sealed unsafe traits are:

1. Traits with no obvious derivable impl:
    * GlobalAlloc
    * Allocator
    * Searcher
    * ReverseSearcher
2. Marker traits
    * TrustedStep
    * TrustedLen
    * DerefPure
    * PinCoerceUnsized
3. TransmuteFrom
4. CloneToUninit

Perhaps `CloneToUninit` is the one most compatible with `#[derive]`:

```rust
#[derive(CloneToUninit)]
struct Packet<Tail: ?Sized> {
    header: Header,
    tail: Tail,
}
```

but the actual impl can satisfy `impl CloneToUninit`'s safety condition without user's intervention, meaning `#[derive(CloneToUninit)]` itself is actually safe.

```rust
unsafe impl<Tail: CloneToUninit + ?Sized> CloneToUninit for Packet<Tail> {
    unsafe fn clone_to_uninit(&self, dst: *mut u8) {
        self.header.clone_to_uninit(dst.add(offset_of_val!(self, header)));
        self.tail.clone_to_uninit(dst.add(offset_of_val!(self, tail)));
    }
}
```

So no I don't have any other examples to support this RFC, i.e. I'm still on the side of the concern "needs fleshed out use case".
